File: frontend/src/lib/api/billing.ts
import api from './client';

export type PlanType = 'free' | 'pro' | 'team';
export type SubscriptionStatus = 'active' | 'canceled' | 'past_due' | 'trialing';

export interface PlanInfo {
	plan: PlanType;
	name: string;
	price_monthly: number;
	tokens_per_month: number;
	features: string[];
	stripe_price_id: string;
}

export interface Subscription {
	id: string;
	user_id: string;
	plan: PlanType;
	status: SubscriptionStatus;
	tokens_used: number;
	tokens_limit: number;
	tokens_remaining: number;
	is_quota_exceeded: boolean;
	stripe_customer_id: string | null;
	stripe_subscription_id: string | null;
}

export const billingApi = {
	getPlans: () =>
		api.get<PlanInfo[]>('/billing/plans'),

	getSubscription: () =>
		api.get<Subscription>('/billing/subscription'),

	createCheckout: (plan: PlanType) =>
		api.post<{ checkout_url: string }>(`/billing/checkout/${plan}`),

	getBillingPortal: () =>
		api.post<{ portal_url: string }>('/billing/portal'),
};

File: backend/app/schemas/billing.py
import uuid
from pydantic import BaseModel
from app.models.subscription import PlanType, SubscriptionStatus


class SubscriptionResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    plan: PlanType
    status: SubscriptionStatus
    tokens_used: int
    tokens_limit: int
    tokens_remaining: int
    is_quota_exceeded: bool
    stripe_customer_id: str | None
    stripe_subscription_id: str | None

    model_config = {"from_attributes": True}


class CheckoutSessionResponse(BaseModel):
    checkout_url: str


class BillingPortalResponse(BaseModel):
    portal_url: str


class PlanInfo(BaseModel):
    plan: PlanType
    name: str
    price_monthly: float
    tokens_per_month: int
    features: list[str]
    stripe_price_id: str
